bin/get_data.js

```diff
@@ -0,0 +1,60 @@
+const fs = require('fs')
+const process = require('process')
+const TOML = require('@iarna/toml')
+
+const header = 'function addEventListener() {}; ';
+const footer = ' exports.experimentPages = experimentPages; ';
+const body = fs.readFileSync('./workers/redirector.js');
+var experimentPages = eval(header + body + footer);
+
+const stage_domain = "https://www.allizom.org"
+const prod_domain = "https://www.mozilla.org"
+var stage_routes = []
+var prod_routes = []
+
+const toml_config_file = fs.readFileSync('./wrangler.toml');
+const toml_obj = TOML.parse(toml_config_file)
+
+// Validate the data
+function pageValidate(page) {
+    // Check all the correct attributes are present
+    if (!('targetPath' in page)) { process.exit(1); }
+    if (!('sandboxPath' in page)) { process.exit(1); }
+    if (!('sampleRate' in page)) { process.exit(1); }
+
+    // Simple validation of our assumption paths start and end with slashes
+    if ( !(page.targetPath.startsWith('/')) ) { process.exit(1); }
+
+    if ( !(page.sandboxPath.startsWith('/')) ) { process.exit(1); }
+
+    // check that they're not the same, as that seems like a common mistake to make
+    if ( page.sandboxPath == page.targetPath ) { process.exit(1); }
+
+    // check the sample rate is between 0 and 1.
+    if (page.sampleRate < 0) { process.exit(1); }
+    if (page.sampleRate > 1) { process.exit(1); }
+}
+
+function buildRoutes(page) {
+    stage_routes.push(stage_domain + page.targetPath + '*')
```

Comment:
I'm a little confused here as to the behavior. The script seems to be building cloudflare worker routes based on the target URLs in `experimentPages`? But I'm not sure this is quite what we want / need, as it would mean deploying different instances of the exact same worker to different and potentially overlapping routes.

A more simple config that may suit our needs is something that allows us to define routes for both stage and prod, independent of what's in `experimentPages`. For example, we may want to publish the worker at the following routes (99.9% of use cases).

- `https://www.allizom.org/en-US/firefox/*`
- `https://www.allizom.org/de/firefox/*`
- `https://www.mozilla.org/en-US/firefox/*`
- `https://www.mozilla.org/de/firefox/*`

The routes would then be a catch-all for any URL that falls under /%LOCALE%/firefox/. The worker script itself is then responsible for looking in `experimentPages` to see if there are any matches.

Happy to talk this through in person it helps?

Reply:
"as it would mean deploying instances of the exact same worker to different and potentially overlapping routes."  I think wrangler works a bit different than what you're saying in the quote. Routes are many to one with the worker.  Each route configured in the wrangler.toml points at the same worker, just catches the traffic and redirects when any match.

They also have a bit of protection, where a worker's route won't be configured (the newest one) if it is identical to an existing worker.  I don't think that it's terribly smart, so it's possible wildcards may not behave correctly (https://www.allizom.org/en-US/* and https://www.allizom.org/en-US/firefox may not be detected as an issue.) but it seems pretty robust.

Thinking about it more I think it makes sense to not use the target path.  Because it seems like you often want to go up one 'directory' and replace with a wildcard. What do you think of adding a new field in the experiement pages definition (workerPath?) to explicitly define the path.  I could then apply dedup the list before passing to the toml.

I think it's really important that the data all live in a single place, because a ton of mistakes have happened to me when data in two places needs to be synced, so trying to use a single source of truth for the worker and for the cloudflare configuration seems good to me.

Reply:
I'm not sure I understand this. Let's say we have the following data:

```javascipt
const experimentPages = [
    {
        'targetPath': `/en-US/firefox/new/`,
        'sandboxPath': `/en-US/exp/firefox/new/`,
        'sampleRate': 0.30
    },
    {
        'targetPath': `/en-US/firefox/`,
        'sandboxPath': `/en-US/exp/firefox/`,
        'sampleRate': 0.30
    }
];
```

Right now, this would result in two routes defined as:

- `https://www.allizom.org/en-US/firefox/new/*`
- `https://www.allizom.org/en-US/firefox/*

I'm not quite clear why we would want / need this - and why we'd need things such as wildcards after /firefox/new/.

Maybe it would help for us to sync up on this, so you can help me understand why wrangler needs to work this way?